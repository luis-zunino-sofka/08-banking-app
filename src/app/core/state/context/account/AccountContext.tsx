import { IGetAllCustomerAccountResponse } from "@core/interfaces";
import { createContext, useContext } from "react";

export interface IAccountContext {
  accounts: IGetAllCustomerAccountResponse[];
  refetchAccounts: () => void;
  balance?: number | null;
}

export const AccountContext = createContext<IAccountContext>({
  accounts: [],
  refetchAccounts: () => {},
  balance: 0,
});

export const useAccountContext = () => useContext(AccountContext);
