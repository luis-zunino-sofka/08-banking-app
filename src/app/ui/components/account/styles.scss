.user-accounts {
  padding: 20px;
  background-color: #f9f9f9;
  border-radius: 8px;

  & p {
    font-size: 1.2rem;
    font-weight: bold;
    margin-bottom: 20px;
    color: #333;
  }

  &__item {
    display: flex;
    gap: 10px;
    justify-content: space-between;
    padding: 10px;
    margin-bottom: 10px;
    background-color: #fff;
    border-radius: 8px;
    box-shadow: 0 2px 5px rgba(0, 0, 0, 0.1);
    transition: transform 0.3s ease;
  }

  &__info {
    font-size: 1rem;
    color: #555;

    & span {
      font-weight: bold;
    }
  }

  .amount {
    color: #4caf50;
  }
}
