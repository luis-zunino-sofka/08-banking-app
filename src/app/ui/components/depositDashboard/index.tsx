import { useDeposit } from "@core/hooks";
import { decryptAES } from "@core/utils";
import { Toaster } from "react-hot-toast";
import { Button } from "../button";

export const DepositDashboard = () => {
  const {
    isLoading,
    accounts,
    errors,
    isAccountSelected,
    handleSubmit,
    register,
    onSubmit,
  } = useDeposit();

  return (
    <div className="deposit">
      <form
        id="deposit-form"
        className="form"
        onSubmit={handleSubmit(onSubmit)}
      >
        <label htmlFor="deposit-source">
          Selecciona el origen del depósito:
        </label>
        <select
          id="deposit-source"
          {...register("depositSource", {
            required: "El monto a depositar es obligatorio",
          })}
        >
          <option value="branch-deposit" defaultValue="branch-deposit">
            Depósito desde sucursal (Sin costo)
          </option>
          <option value="atm">
            Depósito desde cajero automático ($2 de costo)
          </option>
          <option value="account">
            Depósito desde otra cuenta ($1.5 de costo)
          </option>
        </select>
        {errors.depositSource && (
          <span className="error-message">{errors.depositSource.message}</span>
        )}
        <label htmlFor="deposit-amount">Monto:</label>
        <input
          type="number"
          id="deposit-amount"
          placeholder="Monto a depositar"
          required
          min={0}
          {...register("amount", {
            required: "El monto a depositar es obligatorio",
            min: {
              value: 1,
              message: "El monto a depositar debe ser mayor a 0",
            },
          })}
        />
        {errors.amount && (
          <span className="error-message">{errors.amount.message}</span>
        )}
        <div>
          <label htmlFor="accountId">
            {isAccountSelected
              ? "Cuenta de extracción:"
              : "Cuenta a depositar:"}
          </label>
          <select
            id="accountId"
            {...register("accountId", {
              required: "La cuenta bancaria es obligatorio",
            })}
            required
          >
            <option value="">Elige una cuenta</option>
            {accounts.map((a) => (
              <option key={a.accountId} value={a.accountId}>
                Cuenta: {decryptAES(a.encryptedNumber)} - Monto: {a.amount}
              </option>
            ))}
          </select>

          {errors.accountNumberToDeposit && (
            <span className="error-message">
              {errors.accountNumberToDeposit.message}
            </span>
          )}
        </div>
        {isAccountSelected && (
          <>
            <label htmlFor="account-number-to-deposit">
              Selecciona la cuenta a depositar:
            </label>
            <input
              type="text"
              id="account-number-to-deposit"
              placeholder="Cuenta a depositar"
              required
              min={0}
              {...register("accountNumberToDeposit", {
                required: "La cuenta a depositar es obligatorio",
              })}
            />

            {errors.accountNumberToDeposit && (
              <span className="error-message">
                {errors.accountNumberToDeposit.message}
              </span>
            )}
          </>
        )}
        <Button label="Depositar" isLoading={isLoading} />
        <p>* El costo de la transacción se deducirá del saldo.</p>
      </form>
      <Toaster />
    </div>
  );
};
