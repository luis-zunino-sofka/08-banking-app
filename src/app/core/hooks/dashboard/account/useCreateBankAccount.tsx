import { localStorageProperties } from "@core/constants";
import { useAccountContext } from "@core/state/context/account/AccountContext";
import { createCustomerAccount } from "@core/services";
import { useForm } from "react-hook-form";
import toast from "react-hot-toast";
import { ICreateCustomerAccountRequest } from "@core/interfaces";
import { useState } from "react";

interface ICreateBankAccount {
  accountType: string;
  initialBalance: number;
}

export const useCreateBankAccount = () => {
  const { accounts, refetchAccounts } = useAccountContext();
  const [isLoading, setIsLoading] = useState(false);

  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<ICreateBankAccount>();

  const callCreateCustomerAccount = async (
    data: ICreateCustomerAccountRequest
  ) => {
    try {
      setIsLoading(true);
      const result = await createCustomerAccount(data);
      localStorage.setItem(
        localStorageProperties.accountId,
        String(result?.dinBody?.accountId)
      );
      toast("¡Cuenta bancaria creada exitosamente!");
      return result;
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
    } catch (error) {
      toast("Cuentas no cargadas ");
    } finally {
      setIsLoading(false);
    }
  };

  const onSubmit = (data: ICreateBankAccount) => {
    if (data.initialBalance < 0) {
      toast("¡El balance debe de ser mayor a 0!");
      return;
    }
    setIsLoading(true);
    callCreateCustomerAccount({
      customerId: String(
        localStorage.getItem(localStorageProperties.customerId)
      ),
      amount: data.initialBalance,
    });
    refetchAccounts();
  };

  return {
    isLoading,
    accounts,
    errors,
    handleSubmit,
    register,
    onSubmit,
  };
};
