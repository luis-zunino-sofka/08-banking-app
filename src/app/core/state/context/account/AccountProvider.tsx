import { AccountContext } from "./AccountContext";
import { useAccountProvider } from "@core/hooks";

export const AccountProvider = ({
  children,
}: {
  children: React.ReactNode;
}) => {
  const { accounts, state, isRefechingAccounts, refetchAccounts } =
    useAccountProvider();

  return (
    <AccountContext.Provider
      value={{
        accounts,
        balance: state.balance,
        isRefechingAccounts,
        refetchAccounts,
      }}
    >
      {children}
    </AccountContext.Provider>
  );
};
