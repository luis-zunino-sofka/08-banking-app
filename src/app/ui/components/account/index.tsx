import { useCreateBankAccount } from "@core/hooks";
import { decryptAES } from "@core/utils";
import React from "react";
import { Toaster } from "react-hot-toast";
import { Button } from "../button";
import "./styles.scss";

export const CreateBankAccount: React.FC = () => {
  const { isLoading, accounts, errors, handleSubmit, register, onSubmit } =
    useCreateBankAccount();

  return (
    <div>
      <form onSubmit={handleSubmit(onSubmit)} className="form">
        <div style={{ display: "flex", gap: "10px" }}>
          <div>
            <div>
              <label htmlFor="accountType">Tipo de Cuenta</label>
              <select
                id="accountType"
                {...register("accountType", {
                  required: "Selecciona un tipo de cuenta",
                })}
              >
                <option key={1} value="">
                  Selecciona el tipo de cuenta
                </option>
                <option key={2} value="corriente">
                  Corriente
                </option>
              </select>
              <span className="error-message">
                {errors.accountType && errors.accountType.message}
              </span>
            </div>
            <div>
              <label htmlFor="initialBalance">Saldo Inicial</label>
              <input
                type="number"
                id="initialBalance"
                placeholder="Saldo inicial en $"
                min={0}
                {...register("initialBalance", {
                  required: "El saldo inicial es obligatorio",
                  min: {
                    value: 0,
                    message: "El saldo inicial no puede ser negativo",
                  },
                })}
              />
              <span className="error-message">
                {errors.initialBalance && errors.initialBalance.message}
              </span>
            </div>
            <Button label="Crear Cuenta" isLoading={isLoading} />
          </div>
          <div className="user-accounts">
            <p>Cuentas del usuario:</p>
            {accounts.map((a) => (
              <div key={a.accountId} className="user-accounts__item">
                <div className="user-accounts__info">
                  <span>NÂº de cuenta:</span> {decryptAES(a.encryptedNumber)}
                </div>
                <div className="user-accounts__info amount">
                  <span>Monto:</span> $ {a.amount}
                </div>
              </div>
            ))}
          </div>
        </div>
      </form>
      <Toaster />
    </div>
  );
};
