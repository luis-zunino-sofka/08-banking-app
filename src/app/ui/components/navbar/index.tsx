import "./styles.scss";
import { INavItem, IState } from "@core/interfaces";
import { FC } from "react";

export interface IDashboardProps {
  state: IState;
  isRefechingAccounts: boolean;
  navItems: INavItem[];
  refetchAccounts: () => void;
  handleLogout: () => void;
}

export const Navbar: FC<IDashboardProps> = (props) => {
  const {
    state,
    navItems,
    isRefechingAccounts,
    refetchAccounts,
    handleLogout,
  } = props;

  return (
    <nav className="navbar">
      <div className="navbar__title">
        <h1>Bienvenido {state.loginData?.username}</h1>
      </div>
      <div className="navbar__balance">
        <span>Saldo disponible: </span>
        <strong>${state.balance}</strong>
        <button
          disabled={isRefechingAccounts}
          className="navbar__reload"
          onClick={() => refetchAccounts()}
        >
          Actualizar
        </button>
      </div>
      <div className="navbar__nav">
        <ul className="navbar__list">
          {navItems.map((item) => (
            <li key={item.path} className="navbar-item">
              <button
                className={`navbar__link ${
                  item.isActive ? "navbar__link--active" : ""
                }`}
                onClick={item.onClick}
              >
                {item.label}
              </button>
            </li>
          ))}
          <li className="navbar__item">
            <button
              className="navbar__link navbar__logout"
              onClick={handleLogout}
            >
              Cerrar sesi√≥n
            </button>
          </li>
        </ul>
      </div>
    </nav>
  );
};
