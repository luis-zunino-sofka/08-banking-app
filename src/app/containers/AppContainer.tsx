import { useNavbar } from "@core/hooks";
import { useAccountContext, useAppContext } from "@core/state";
import { Navbar } from "@ui/components/navbar";
import { AppLayout } from "@ui/layouts";
import React from "react";
import { Toaster } from "react-hot-toast";

export const AppContainer = ({ children }: { children: React.ReactNode }) => {
  const { state } = useAppContext();
  const { refetchAccounts } = useAccountContext();
  const { navItems, handleLogout } = useNavbar();

  return (
    <AppLayout>
      <Navbar
        state={state}
        handleLogout={handleLogout}
        refetchAccounts={refetchAccounts}
        navItems={navItems}
      />
      {children}
      <Toaster />
    </AppLayout>
  );
};
