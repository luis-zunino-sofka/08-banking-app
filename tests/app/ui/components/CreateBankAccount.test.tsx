import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { afterEach, beforeEach, describe, expect, test, vi } from "vitest";
import React from "react";
import { CreateBankAccount } from "../../../../src/app/ui/components/account";
import { AccountContextMock } from "../containers/AppContainer.test";
import "@testing-library/jest-dom";
import { mockAccountState, mockAppState } from "../../../mocks/dataMocked";

vi.mock("@core/services", () => ({
  createCustomerAccount: vi.fn(),
}));

vi.mock("@core/state/context/account/AccountContext", () => ({
  useAccountContext: vi.fn(),
}));

vi.mock("react-hook-form", () => ({
  useForm: () => ({
    register: vi.fn(),
    formState: {
      errors: {
        name: { message: "El nombre es obligatorio" },
        lastName: { message: "El apellido es obligatorio" },
      },
    },
    handleSubmit: vi.fn(),
  }),
}));

vi.mock("./hooks", () => ({
  useCreateBankAccount: vi.fn(),
}));

global.localStorage.setItem = vi.fn();
global.localStorage.getItem = vi.fn(() => "1234");

const Renderear = () => {
  return (
    <AccountContextMock
      mockAppState={mockAppState}
      children={<CreateBankAccount />}
      mockAccountState={mockAccountState}
    />
  );
};

describe("CreateBankAccount", () => {
  const mockOnSubmit = vi.fn();

  beforeEach(() => {
    vi.mock("@core/hooks", () => ({
      useCreateBankAccount: () => ({
        isLoading: false,
        accounts: [
          { accountId: "1", encryptedNumber: "1234", amount: 1000 },
          { accountId: "2", encryptedNumber: "5678", amount: 2500 },
        ],
        errors: {},
        handleSubmit: vi.fn((fn) => fn),
        register: vi.fn(),
        onSubmit: mockOnSubmit,
      }),
    }));
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  test("Match snapshoot", () => {
    const { asFragment } = render(<Renderear />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("renderiza el formulario correctamente", () => {
    render(<Renderear />);

    screen.getByLabelText(/Tipo de Cuenta/i);
    screen.getByLabelText(/Saldo Inicial/i);
    screen.getByText(/Crear Cuenta/i);
    screen.getByText(/Tus cuentas:/i);
    screen.getAllByText(/Nº de cuenta:/i);
    screen.getAllByText(/Monto:/i);
    screen.getByText(/2500/i);
  });

  test("llama a la función onSubmit correctamente cuando el formulario es enviado", async () => {
    render(<Renderear />);

    const accountTypeSelect = screen.getByLabelText(/Tipo de Cuenta/i);
    const initialBalanceInput = screen.getByLabelText(/Saldo Inicial/i);
    const submitButton = screen.getByRole("button", {
      name: /^Crear Cuenta$/i,
    });
    waitFor(() => {
      fireEvent.change(accountTypeSelect, { target: { value: "corriente" } });
      fireEvent.change(initialBalanceInput, { target: { value: 1000 } });
      fireEvent.submit(submitButton);
    });

    expect(accountTypeSelect).toHaveValue("corriente");
    expect(initialBalanceInput).toHaveValue(1000);

    waitFor(() => expect(mockOnSubmit).toHaveBeenCalled());
  });

  test("llama a la función cuando el formulario es enviado correctamente", async () => {
    render(<Renderear />);

    const submitButton = screen.getByRole("button", {
      name: /^Crear Cuenta$/i,
    });
    const accountTypeSelect = screen.getByLabelText(/Tipo de Cuenta/i);
    const initialBalanceInput = screen.getByLabelText(/Saldo Inicial/i);

    waitFor(() => {
      fireEvent.change(accountTypeSelect, { target: { value: "corriente" } });
      fireEvent.change(initialBalanceInput, { target: { value: "1000" } });
      fireEvent.submit(submitButton);
    });

    expect(submitButton).toHaveAttribute("type", "submit");

    expect(accountTypeSelect).toHaveValue("corriente");
    expect(initialBalanceInput).toHaveValue(1000);

    waitFor(() => {
      expect(mockOnSubmit).toHaveBeenCalledWith({
        customerId: "1234",
        amount: 1000,
      });
    });
  });
});
