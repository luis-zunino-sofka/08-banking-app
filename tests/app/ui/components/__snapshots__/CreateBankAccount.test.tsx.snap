// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AccountProvider > Match snapshoot 1`] = `<DocumentFragment />`;

exports[`CreateBankAccount > Match snapshoot 1`] = `
<DocumentFragment>
  <div>
    <form
      class="form"
    >
      <div
        style="display: flex; gap: 10px;"
      >
        <div>
          <div>
            <label
              for="accountType"
            >
              Tipo de Cuenta
            </label>
            <select
              id="accountType"
            >
              <option
                value=""
              >
                Selecciona el tipo de cuenta
              </option>
              <option
                value="corriente"
              >
                Corriente
              </option>
            </select>
            <span
              class="error-message"
            />
          </div>
          <div>
            <label
              for="initialBalance"
            >
              Saldo Inicial
            </label>
            <input
              id="initialBalance"
              min="0"
              placeholder="Saldo inicial en $"
              type="number"
            />
            <span
              class="error-message"
            />
          </div>
          <button
            class="button"
            type="submit"
          >
            Crear Cuenta
          </button>
        </div>
        <div
          class="user-accounts"
        >
          <p>
            Tus cuentas:
          </p>
          <div
            class="user-accounts__item"
          >
            <div
              class="user-accounts__info"
            >
              <span>
                Nº de cuenta:
              </span>
               encrypted1
            </div>
            <div
              class="user-accounts__info amount"
            >
              <span>
                Monto:
              </span>
               $ 1000
            </div>
          </div>
          <div
            class="user-accounts__item"
          >
            <div
              class="user-accounts__info"
            >
              <span>
                Nº de cuenta:
              </span>
               encrypted2
            </div>
            <div
              class="user-accounts__info amount"
            >
              <span>
                Monto:
              </span>
               $ 2500
            </div>
          </div>
        </div>
      </div>
    </form>
    <div
      style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
    />
  </div>
</DocumentFragment>
`;
